File: backend/app/routers/webhooks.py
import base64
import json
from fastapi import APIRouter, Request
from app.errors import bad_request
from app.supabase_rest import select
from app.crypto import decrypt
from app.services import gmail, outlook
from app.services.email import parse_gmail_message, parse_outlook_message
from app.services.store import store_email
from app.queue import enqueue_job
from app.services.cache import bump_user_cache_version
from app.webhook_security import verify_signature

router = APIRouter(prefix="/webhooks", tags=["webhooks"])


@router.post("/gmail")
async def gmail_webhook(request: Request):
    body = await request.body()
    verify_signature(body, request.headers.get("x-unisync-signature"))
    try:
        payload = json.loads(body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        bad_request("Invalid Gmail webhook payload")
    message = payload.get("message", {})
    data = message.get("data")
    if not data:
        return {"status": "ok"}

    try:
        decoded = base64.b64decode(data).decode("utf-8")
        event = json.loads(decoded)
    except (ValueError, UnicodeDecodeError, json.JSONDecodeError):
        bad_request("Invalid Gmail webhook event")
    email_address = event.get("emailAddress")
    history_id = event.get("historyId")
    if not email_address or not history_id:
        return {"status": "ok"}

    accounts = await select(
        "linked_accounts",
        "id,user_id,access_token_enc",
        filters=[("provider", "eq", "gmail"), ("email_address", "eq", email_address)],
        use_service=True,
    )
    if not accounts:
        return {"status": "ok"}
    account = accounts[0]

    access_token = decrypt(account["access_token_enc"])
    history = await gmail.fetch_history(access_token, history_id)
    history_items = history.get("history", []) or []
    message_ids = []
    for item in history_items:
        for added in item.get("messagesAdded", []) or []:
            msg = added.get("message", {})
            if msg.get("id"):
                message_ids.append(msg["id"])

    for msg_id in message_ids:
        raw = await gmail.fetch_message(access_token, msg_id)
        parsed = await parse_gmail_message(
            raw,
            lambda attachment_id, mid=msg_id: gmail.fetch_attachment(access_token, mid, attachment_id),
        )
        parsed["provider"] = "gmail"
        email_id = await store_email(account["user_id"], account["id"], parsed)
        await enqueue_job("process_email", email_id)
        await bump_user_cache_version(account["user_id"])

    return {"status": "ok"}


@router.post("/outlook")
async def outlook_webhook(request: Request):
    body = await request.body()
    verify_signature(body, request.headers.get("x-unisync-signature"))
    try:
        payload = json.loads(body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        bad_request("Invalid Outlook webhook payload")
    notifications = payload.get("value", []) or []

    for note in notifications:
        if note.get("clientState") and note.get("clientState") != "unisync":
            continue
        subscription_id = note.get("subscriptionId")
        if not subscription_id:
            continue
        resource_data = note.get("resourceData", {})
        message_id = resource_data.get("id")
        if not message_id:
            continue

        accounts = await select(
            "linked_accounts",
            "id,user_id,access_token_enc",
            filters=[("provider", "eq", "outlook"), ("subscription_id", "eq", subscription_id)],
            use_service=True,
        )
        if not accounts:
            continue
        account = accounts[0]

        access_token = decrypt(account["access_token_enc"])
        raw = await outlook.fetch_message(access_token, message_id)
        parsed = parse_outlook_message(raw)
        parsed["provider"] = "outlook"
        email_id = await store_email(account["user_id"], account["id"], parsed)
        await enqueue_job("process_email", email_id)
        await bump_user_cache_version(account["user_id"])

    return {"status": "ok"}

File: backend2/src/routes/webhooks.ts
import { Router } from "express";

import { decrypt, encrypt } from "../crypto.js";
import { badRequest } from "../errors.js";
import { enqueueJob } from "../queue.js";
import { bumpUserCacheVersion } from "../services/cache.js";
import { fetchAttachment, fetchHistory, fetchMessage as fetchGmailMessage, refreshToken as refreshGmailToken } from "../services/gmail.js";
import { parseGmailMessage, parseOutlookMessage } from "../services/email.js";
import { fetchMessage as fetchOutlookMessage } from "../services/outlook.js";
import { storeEmail } from "../services/store.js";
import { select, update } from "../supabaseRest.js";
import { verifySignature } from "../webhookSecurity.js";

const router = Router();

async function withFreshGmailAccessToken(account: Record<string, any>, fn: (accessToken: string) => Promise<any>) {
  let accessToken = decrypt(account.access_token_enc);
  const refreshToken = decrypt(account.refresh_token_enc || "");
  try {
    return await fn(accessToken);
  } catch (error: any) {
    if (![401, 403].includes(Number(error?.status)) || !refreshToken) throw error;
    const refreshed = await refreshGmailToken(refreshToken);
    accessToken = refreshed.access_token || "";
    if (!accessToken) badRequest("Gmail token expired. Please reconnect Gmail.");
    await update("linked_accounts", { access_token_enc: encrypt(accessToken) }, {
      filters: [["id", "eq", account.id]],
      useService: true
    });
    return fn(accessToken);
  }
}

router.post("/gmail", async (request, response, next) => {
  try {
    const bodyBuffer = Buffer.isBuffer(request.body) ? request.body : Buffer.from(JSON.stringify(request.body ?? {}));
    verifySignature(bodyBuffer, request.header("x-unisync-signature"));
    const payload = typeof request.body === "object" ? request.body : JSON.parse(bodyBuffer.toString("utf8"));
    const data = payload?.message?.data;
    if (!data) {
      response.json({ status: "ok" });
      return;
    }
    const event = JSON.parse(Buffer.from(data, "base64").toString("utf8"));
    if (!event.emailAddress || !event.historyId) {
      response.json({ status: "ok" });
      return;
    }
    const accounts = await select("linked_accounts", "id,user_id,access_token_enc,refresh_token_enc", {
      filters: [["provider", "eq", "gmail"], ["email_address", "eq", event.emailAddress]],
      useService: true
    });
    if (!accounts.length) {
      response.json({ status: "ok" });
      return;
    }
    const account = accounts[0];
    await withFreshGmailAccessToken(account, async (accessToken) => {
      const history = await fetchHistory(accessToken, event.historyId);
      for (const item of history.history ?? []) {
        for (const added of item.messagesAdded ?? []) {
          const messageId = added?.message?.id;
          if (!messageId) continue;
          const raw = await fetchGmailMessage(accessToken, messageId);
          const parsed = await parseGmailMessage(raw, (attachmentId) => fetchAttachment(accessToken, messageId, attachmentId));
          const emailId = await storeEmail(account.user_id, account.id, { ...parsed, provider: "gmail" });
          await enqueueJob("process_email", emailId);
          await bumpUserCacheVersion(account.user_id);
        }
      }
    });
    response.json({ status: "ok" });
  } catch (error) {
    next(error);
  }
});

router.post("/outlook", async (request, response, next) => {
  try {
    const bodyBuffer = Buffer.isBuffer(request.body) ? request.body : Buffer.from(JSON.stringify(request.body ?? {}));
    verifySignature(bodyBuffer, request.header("x-unisync-signature"));
    const payload = typeof request.body === "object" ? request.body : JSON.parse(bodyBuffer.toString("utf8"));
    for (const note of payload?.value ?? []) {
      if (note.clientState && note.clientState !== "unisync") continue;
      if (!note.subscriptionId || !note.resourceData?.id) continue;
      const accounts = await select("linked_accounts", "id,user_id,access_token_enc", {
        filters: [["provider", "eq", "outlook"], ["subscription_id", "eq", note.subscriptionId]],
        useService: true
      });
      if (!accounts.length) continue;
      const account = accounts[0];
      const accessToken = decrypt(account.access_token_enc);
      const raw = await fetchOutlookMessage(accessToken, note.resourceData.id);
      const emailId = await storeEmail(account.user_id, account.id, { ...parseOutlookMessage(raw), provider: "outlook" });
      await enqueueJob("process_email", emailId);
      await bumpUserCacheVersion(account.user_id);
    }
    response.json({ status: "ok" });
  } catch (error) {
    next(error);
  }
});

export default router;
